README.md

```diff
@@ -1,10 +1,239 @@
+
+### Concurrency
+To reveal the true power of asynchronous programming, we have to introduce *concurrency* in our program.
+If our goal is to send only one HTTP request and wait for it,
+there is no gain to deal with an asynchronous request.
+However, as soon as you have at least two goals to reach,
+asynchronous functions will improve your performances thanks to concurrency.
+To resolve several independent [`Promises`](src/Promise.php),
+use [`EventLoop::promiseAll`](src/EventLoop.php) method to create a new [`Promise`](src/Promise.php)
+that will be resolved when all others are resolved.
+```php
+<?php
+
+use M6Web\Tornado\EventLoop;
+use M6Web\Tornado\HttpClient;
+use Psr\Http\Message\RequestInterface;
+
+function mySynchronousAction(EventLoop $eventLoop, HttpClient $httpClient, array $requests)
+{
+    $myPromises = [];
+    /** @var RequestInterface $request */
+    foreach ($requests as $request) {
+        $myPromises[] = getJsonResponse($httpClient, $request);
+    }
+    
+    $allJsonArrays = $eventLoop->wait(
+        $eventLoop->promiseAll(...$myPromises)
+    );
+    
+    var_dump($allJsonArrays);
+}
+```
+
+It's important to note that 
+it will be more efficient to use [`EventLoop::promiseAll`](src/EventLoop.php)
+instead of waiting each input [`Promise`](src/Promise.php) consecutively,
+because of concurrency.
+Each time that you have several promises to resolve,
+ask yourself if you could wait them concurrently, especially when you deal with `foreach` loops. 
+
+### Resolving your own promises
+By design, you cannot resolve a promise by yourself, you will need a [`Deferred`](src/Deferred.php).
+It allows you to create a [`Promise`](src/Promise.php) and to resolve (or reject) it
+while not exposing these advanced controls.
+```php
+<?php
+
+use M6Web\Tornado\Deferred;
+use M6Web\Tornado\EventLoop;
+use M6Web\Tornado\HttpClient;
+use M6Web\Tornado\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+function loopUntilStatus418(EventLoop $eventLoop, HttpClient $httpClient, RequestInterface $request, Deferred $deferred)
+{
+    do {
+        yield $eventLoop->delay(1000 /*ms*/);
+        /** @var ResponseInterface $response */
+        $response = yield $httpClient->sendRequest($request);
+    } while ($response->getStatusCode() !== 418);
+
+    $deferred->resolve($response->getBody());
+}
+
+function get418Response(EventLoop $eventLoop, HttpClient $httpClient, RequestInterface $request): Promise
+{
+    $deferred = $eventLoop->deferred();
+    $eventLoop->async(loopUntilStatus418($eventLoop, $httpClient, $request, $deferred));
+    
+    return $deferred->getPromise();
+}
+
+function mySynchronousAction(EventLoop $eventLoop, HttpClient $httpClient, RequestInterface $request)
+{
+    /** @var ResponseInterface $response */
+    $response = $eventLoop->wait(get418Response($eventLoop, $httpClient, $request));
+
+    echo (string)$response->getBody();
```

Comment:
you already resolve the response body, here you try to access the body again, is it correct?

Reply:
You are totally right… I should execute these samples in a dedicated example 🤔